File: packages/runner/src/executor/http.ts
import { readFile } from "node:fs/promises";
import { resolve as resolvePath } from "node:path";

import { renderDeep } from "@agentic-test-hub/plugin";
import type { Operation } from "@agentic-test-hub/plugin";

import {
  ExecutorError,
  type ExecutionContext,
  type ExecutionResult,
  type Executor,
} from "./types.ts";

/** How a request is made. Injected so tests need no server. */
export type FetchFn = typeof fetch;

function joinUrl(base: string, path: string): string {
  return `${base.replace(/\/+$/, "")}/${path.replace(/^\/+/, "")}`;
}

/**
 * Reads a response body without deciding in advance what it is.
 *
 * Content type is advisory, and services under test get it wrong — an error
 * page served as JSON, a body declared JSON that is empty. Text is always
 * kept so that evidence and assertions have something to work with even when
 * parsing fails.
 */
async function readBody(response: Response): Promise<{ body: unknown; text: string }> {
  const text = await response.text();
  if (text === "") return { body: undefined, text };
  try {
    return { body: JSON.parse(text), text };
  } catch {
    return { body: text, text };
  }
}

/** Runs `http` operations. */
export class HttpExecutor implements Executor<"http"> {
  readonly kind = "http" as const;
  private readonly fetchFn: FetchFn;

  /**
   * @param fetchFn - How to make a request. Defaults to the platform's
   *   `fetch`; tests supply a stand-in.
   */
  constructor(fetchFn: FetchFn = fetch) {
    this.fetchFn = fetchFn;
  }

  async run(
    operation: Extract<Operation, { executor: "http" }>,
    context: ExecutionContext,
  ): Promise<ExecutionResult> {
    const connection = context.manifest.connections[operation.connection];
    if (connection?.kind !== "http") {
      throw new ExecutorError(
        `connection "${operation.connection}" is not an http connection`,
        operation.connection,
      );
    }

    const rendered = renderDeep(
      {
        baseUrl: connection.baseUrl,
        path: operation.path,
        headers: { ...connection.headers, ...operation.headers },
        body: operation.body,
      },
      context.scopes,
    );

    let payload: string | undefined;
    if (operation.bodyFile !== undefined) {
      payload = await readFile(resolvePath(context.root, operation.bodyFile), "utf8");
    } else if (rendered.body !== undefined) {
      payload = typeof rendered.body === "string" ? rendered.body : JSON.stringify(rendered.body);
    }

    const url = joinUrl(rendered.baseUrl, rendered.path);
    const headers = { ...rendered.headers };
    if (payload !== undefined && headers["Content-Type"] === undefined) {
      headers["Content-Type"] = "application/json";
    }

    const timeout = AbortSignal.timeout(operation.timeoutMs);
    const signal =
      context.signal === undefined ? timeout : AbortSignal.any([timeout, context.signal]);

    const started = Date.now();
    try {
      const response = await this.fetchFn(url, {
        method: operation.method,
        headers,
        signal,
        ...(payload === undefined ? {} : { body: payload }),
      });
      const { body, text } = await readBody(response);
      return {
        operation: `${operation.method} ${url}`,
        // Any answer from the service is a completed operation, including a
        // rejection. Which statuses are acceptable is the case's business.
        ok: true,
        durationMs: Date.now() - started,
        status: response.status,
        body,
        stdout: text,
      };
    } catch (cause) {
      return {
        operation: `${operation.method} ${url}`,
        ok: false,
        durationMs: Date.now() - started,
        failure: `request did not complete: ${cause instanceof Error ? cause.message : String(cause)}`,
      };
    }
  }
}

File: python/agentic-test-hub-runner/src/agentic_test_hub_runner/executors/http.py
"""Runs `http` operations, mirroring packages/runner/src/executor/http.ts."""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any

import httpx

from ..template import render_deep
from ..types import ExecutionContext, ExecutionResult, ExecutorError


class HttpExecutor:
    """Runs `http` operations.

    @param client: An `httpx.Client` to send requests through. Defaults to a
        real client; tests pass one built on `httpx.MockTransport`.
    """

    kind = "http"

    def __init__(self, client: httpx.Client | None = None) -> None:
        self._client = client or httpx.Client()

    def run(self, operation: dict[str, Any], context: ExecutionContext) -> ExecutionResult:
        connection = context.manifest.connections.get(operation["connection"])
        if connection is None or connection.get("kind") != "http":
            raise ExecutorError(
                f'connection "{operation["connection"]}" is not an http connection',
                operation["connection"],
            )

        rendered = render_deep(
            {
                "baseUrl": connection["baseUrl"],
                "path": operation["path"],
                "headers": {**connection.get("headers", {}), **operation.get("headers", {})},
                "body": operation.get("body"),
            },
            context.scopes,
        )

        payload: str | None = None
        body_file = operation.get("bodyFile")
        body_param = operation.get("bodyParam")
        if body_file is not None:
            payload = (Path(context.root) / body_file).read_text("utf8")
        elif body_param is not None:
            # Uses the param's value directly, unrendered - render_deep only
            # ever substitutes into text (see template.py's render), which
            # cannot express splicing a whole structured value into a body
            # position. Mirrors packages/runner/src/executor/http.ts.
            params = context.scopes.get("param") or {}
            if body_param not in params:
                raise ExecutorError(
                    f'operation needs param "{body_param}" for its body', operation["connection"]
                )
            payload = json.dumps(params[body_param])
        elif rendered["body"] is not None:
            body = rendered["body"]
            payload = body if isinstance(body, str) else json.dumps(body)

        method = operation.get("method", "GET")
        url = f"{rendered['baseUrl'].rstrip('/')}/{rendered['path'].lstrip('/')}"
        headers = dict(rendered["headers"])
        if payload is not None and "Content-Type" not in headers:
            headers["Content-Type"] = "application/json"

        started = time.monotonic()
        try:
            response = self._client.request(
                method,
                url,
                headers=headers,
                content=payload,
                timeout=operation.get("timeoutMs", 60_000) / 1000,
            )
            text = response.text
            try:
                body_value: Any = response.json() if text else None
            except ValueError:
                body_value = text or None
            return ExecutionResult(
                operation=f"{method} {url}",
                ok=True,
                duration_ms=(time.monotonic() - started) * 1000,
                status=response.status_code,
                body=body_value,
                stdout=text,
            )
        except httpx.HTTPError as cause:
            return ExecutionResult(
                operation=f"{method} {url}",
                ok=False,
                duration_ms=(time.monotonic() - started) * 1000,
                failure=f"request did not complete: {cause}",
            )
